src/guiapi/src/window_frame.cpp

```diff
@@ -1,92 +1,223 @@
 #include "sound.hpp"
+#include "ScreenHandler.hpp"
 
-void window_frame_init(window_frame_t *window) {
-    if (rect_empty_ui16(window->rect)) //use display rect curent is empty
-        window->rect = rect_ui16(0, 0, display::GetW(), display::GetH());
-    window->flg |= WINDOW_FLG_ENABLED | WINDOW_FLG_PARENT;
-    window->color_back = COLOR_BLACK;
+window_frame_t::window_frame_t(window_t *parent, rect_ui16_t rect, is_dialog_t dialog)
+    : window_t(parent, rect, dialog)
+    , first(nullptr)
+    , last(nullptr) {
+    Enable();
+    color_back = COLOR_BLACK;
 }
 
-void window_frame_done(window_frame_t *window) {
+//register sub win
+void window_frame_t::RegisterSubWin(window_t *win) {
+    if (first && last) {
+        if (win->IsDialog()) {
+            window_t *pWin = first;
+            while (pWin) {
+                pWin->Hide(); //todo check dialog intersection
+                //define shadowed by dialog flag
+                pWin = pWin->GetNext();
+            }
+        }
+        last->SetNext(win);
+        last = last->GetNext();
+    } else {
+        first = last = win;
+    }
 }
 
-void window_frame_draw(window_frame_t *window) {
-    if (window->f_visible) {
-        if (window->f_invalid) {
-            rect_ui16_t rc = window->rect;
-            display::FillRect(rc, window->color_back);
-            window->f_invalid = 0;
-            window_invalidate_children(window->id);
+//unregister sub win
+void window_frame_t::UnregisterSubWin(window_t *win) {
+    window_t *prev = GetPrevSubWin(win);
+    if (prev) {
+        prev->SetNext(win->GetNext());
+        if (win == last)
+            last = prev;
+    }
+
+    if (last->IsDialog()) {
+        //todo check intersection with this dialog
+        //and show windows that have no intersection with it
+        last->Show();
+    } else {
+        //show all windows
+        //todo check shadowed by dialog flag
+        window_t *pWin = first;
+        while (pWin) {
+            pWin->Show();
+            pWin = pWin->GetNext();
         }
-        window_draw_children(window->id);
+
+        //todo remove after menu refactoring - menu items must be windows
+        //needed for menu
+        Invalidate();
     }
 }
 
-void window_frame_event(window_frame_t *window, uint8_t event, void *param) {
-    int16_t id;
-    int dif;
+window_t *window_frame_t::GetFirst() const {
+    return first;
+}
+
+window_t *window_frame_t::GetLast() const {
+    return last;
+}
+
+void window_frame_t::draw() {
+    if (!IsVisible())
+        return;
+    bool setChildernInvalid = false;
+
+    if (IsInvalid()) {
+        unconditionalDraw();
+        Validate();
+        setChildernInvalid = true;
+    }
+
+    window_t *ptr = first;
+    while (ptr) {
+        if (setChildernInvalid)
+            ptr->Invalidate();
+        ptr->Draw();
+        ptr = ptr->GetNext();
+    }
+}
+
+void window_frame_t::windowEvent(window_t *sender, uint8_t event, void *param) {
+    int dif = (int)param;
```

Comment:
(int)*param?

Reply:
it is old code, to be removed.
there will be no void* parameter in future